btree.go

```diff
@@ -0,0 +1,346 @@
+	// b-trees are inconsistently categorized either by the number of children
+	// or by the number of keys. We choose the former.
+	v int
+}
+
+func newBtree(opts btreeOpts) *btree {
+	return &btree{&leaf{}, opts}
+}
+
+func (bt *btree) insert(ng ngram) {
+	if leftNode, rightNode, newKey, ok := bt.root.maybeSplit(bt.opts); ok {
+		bt.root = &innerNode{keys: []ngram{newKey}, children: []node{leftNode, rightNode}}
+	}
+	bt.root.insert(ng, bt.opts)
+}
+
+// find returns the tuple (bucketIndex, postingIndexOffset), both of which are
+// stored at the leaf level. They are effectively pointers to the bucket and
+// the posting lists for ngrams stored in the bucket. Since ngrams and their
+// posting lists are stored in order, knowing the index of the posting list of
+// the first item in the bucket is sufficient.
+func (bt *btree) find(ng ngram) (int, int) {
+	if bt.root == nil {
+		return -1, -1
+	}
+	return bt.root.find(ng)
+}
+
+func (bt *btree) visit(f func(n node)) {
+	bt.root.visit(f)
+}
+
+func (bt *btree) sizeBytes() int {
+	sz := 2 * 8 // opts
+
+	bt.visit(func(n node) {
+		sz += n.sizeBytes()
+	})
+
+	return sz
+}
+
+type node interface {
+	insert(ng ngram, opts btreeOpts)
+	maybeSplit(opts btreeOpts) (left node, right node, newKey ngram, ok bool)
+	find(ng ngram) (int, int)
+	visit(func(n node))
+	sizeBytes() int
+}
+
+type innerNode struct {
+	keys     []ngram
+	children []node
+}
+
+type leaf struct {
+	bucketIndex int
+	// postingIndexOffset is the index of the posting list of the first ngram
+	// in the bucket. This is enough to determine the index of the posting list
+	// for every other key in the bucket.
+	postingIndexOffset int
+
+	// Optimization: Because we insert ngrams in order, we don't actually have
+	// to fill the buckets. We just have to keep track of the size of the
+	// bucket, so we know when to split, and the key that we have to propagate
+	// up to the parent node when we split.
+	//
+	// If in the future we decide to mutate buckets, we have to replace
+	// bucketSize and splitKey by []ngram.
+	bucketSize int
+	splitKey   ngram
+}
+
+func (n *innerNode) sizeBytes() int {
+	return len(n.keys)*ngramEncoding + len(n.children)*int(interfaceBytes)
+}
+
+func (n *leaf) sizeBytes() int {
+	return 4 * 8
+}
+
+func (n *leaf) insert(ng ngram, opts btreeOpts) {
+	n.bucketSize++
+
+	if n.bucketSize == (opts.bucketSize/2)+1 {
+		n.splitKey = ng
+	}
+}
+
+func (n *innerNode) insert(ng ngram, opts btreeOpts) {
+	insertAt := func(i int) {
+		// Invariant: Nodes always have a free slot.
+		//
+		// We split full nodes on the the way down to the leaf. This has the
+		// advantage that inserts are handled in a single pass.
+		if leftNode, rightNode, newKey, ok := n.children[i].maybeSplit(opts); ok {
+			n.keys = append(n.keys[0:i], append([]ngram{newKey}, n.keys[i:]...)...)
+			n.children = append(n.children[0:i], append([]node{leftNode, rightNode}, n.children[i+1:]...)...)
```

Comment:
very minor: likely a very minor optimization, but I wonder if you add a method to innerNode like `shrink` which re-allocates these structs such that cap == len. You run it after constructing and we may get some memory savings that _might_ be noticeable w.r.t. the btree.

Reply:
I believe I already do what you propose in innerNode.maybeSplit. Once we split a node we never touch the left node again and I set cap = len there. See lines 206ff.